src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1085 @@
+    } else if (cur_cost + adaptive_beam < cur_cutoff) {
+      cur_cutoff = cur_cost + adaptive_beam; // a tighter boundary
+    }
+    // If "tok" has any existing forward links, delete them,
+    // because we're about to regenerate them.  This is a kind
+    // of non-optimality (remember, this is the simple decoder),
+    DeleteForwardLinks(tok);  // necessary when re-visiting
+    for (fst::ArcIterator<FST> aiter(*fst_, state);
+         !aiter.Done();
+         aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      bool changed;
+      if (arc.ilabel == 0) {  // propagate nonemitting
+        BaseFloat graph_cost = arc.weight.Value();
+        BaseFloat tot_cost = cur_cost + graph_cost;
+        if (tot_cost < cur_cutoff) {
+          Token *new_tok = FindOrAddToken(arc.nextstate, frame, tot_cost,
+                                          tok, &prev_toks_, &changed);
+
+          // Add ForwardLink from tok to new_tok. Put it on the head of
+          // tok->link list
+          tok->links = new ForwardLinkT(new_tok, 0, arc.olabel,
+                                        graph_cost, 0, tok->links);
+          
+          // "changed" tells us whether the new token has a different
+          // cost from before, or is new.
+          if (changed) {
+            cur_queue_.Push(new_tok);
+          }
+        }
+      } else {  // propagate emitting
+        BaseFloat graph_cost = arc.weight.Value(),
+                  ac_cost = cost_offset - decodable->LogLikelihood(frame, arc.ilabel),
+                  cur_cost = tok->tot_cost,
+                  tot_cost = cur_cost + ac_cost + graph_cost;
+        if (tot_cost > next_cutoff) continue;
+        else if (tot_cost + adaptive_beam < next_cutoff) {
+          next_cutoff = tot_cost + adaptive_beam;  // a tighter boundary for emitting
+        }
+
+        // no change flag is needed
+        Token *next_tok = FindOrAddToken(arc.nextstate, frame + 1, tot_cost,
+                                         tok, &cur_toks_, NULL);
+        // Add ForwardLink from tok to next_tok. Put it on the head of tok->link
+        // list
+        tok->links = new ForwardLinkT(next_tok, arc.ilabel, arc.olabel,
+                                      graph_cost, ac_cost, tok->links);
+      }
+    }  // for all arcs
+  }  // end of while loop
+
+  // Store the offset on the acoustic likelihoods that we're applying.
+  // Could just do cost_offsets_.push_back(cost_offset), but we
+  // do it this way as it's more robust to future code changes.
+  // Set the cost_offset_ for next frame, it equals "- best_cost_on_next_frame".
+  cost_offsets_.resize(frame + 2, 0.0);
+  cost_offsets_[frame + 1] = adaptive_beam - next_cutoff;
+
+  {  // This block updates adaptive_beam_
+    BaseFloat beam_used_this_frame = adaptive_beam;
+    Token *tok = cur_queue_.Pop();
+    if (tok != NULL) {
+      // The queue would only be nonempty if we hit the max-active constraint.
```

Comment:
actuall, change  this comment to 
```
// We hit the max-active contraint, meaning we effectively pruned to a beam
// tighter than `beam`.  Work out what this was, it will be used to update
// `adaptive_beam`.
```